dev => main

dlomanov opened the pull request:
Сервер готов, клиент активно допиливается

dlomanov committed:
add user and entry entities

dlomanov committed:
add user use-case (sign-in, sign-up)

dlomanov committed:
add db and user repo

dlomanov committed:
add grpc for user use-case

dlomanov committed:
add server infrastructure test

dlomanov committed:
add entry use case with repo

dlomanov committed:
add auth middleware

dlomanov committed:
rework entry use-case

dlomanov committed:
add grpc entry-service with tests

dlomanov committed:
add data encryption

dlomanov committed:
support GRPC-TLS

dlomanov committed:
add versioning

dlomanov committed:
add force update method for entries

dlomanov committed:
resolve entry version conflict by saving both version

dlomanov committed:
add entry's client-server difference definition

dlomanov committed:
raw client TUI

dlomanov committed:
write client logs to file

dlomanov committed:
separate entries into client and server version

dlomanov committed:
add local storage for client

- using sqlite as local storage
- client ui refactoring
- client looking for config nearby
- added raw sync mechanism
- tests

dlomanov committed:
sync client with server when possible

bambruysk commented on cmd/server/config/config.go:
```diff
@@ -59,39 +70,79 @@ func (c *config) readConfig() {
+func (c *config) toConfig() *srvcfg.Config {
+	cert, certKey := c.readCert()
+
+	res := &srvcfg.Config{
```
можно просто return тут

dlomanov replied:
пофиксил

bambruysk commented on cmd/server/config/config.go:
```diff
@@ -59,39 +70,79 @@ func (c *config) readConfig() {
+
+	cert, err := os.ReadFile(c.CertPath)
+	if err != nil {
+		log.Fatalf("failed to read TLS-certificate: %v", err)
```
лучше не падать тут - а вернуть ошибку - тогда такую функцию проще обложить тестами

dlomanov replied:
вернул ошибку, падаю в функции выше

bambruysk commented on internal/apps/client/infra/deps/container.go:
```diff
@@ -0,0 +1,123 @@
+	defer cancel()
+
+	if err := c.Conn.Close(); err != nil {
+		merr = multierr.Append(merr, fmt.Errorf("container: failed to close GRPC-connection: %w", err))
```
multierr выходит из моды - его функционал сейчас есть в стандартном errors.Join

dlomanov replied:
заменил multierr на errors

bambruysk commented on internal/apps/client/migrations/migrations.go:
```diff
@@ -0,0 +1,38 @@
+	for i, f := range files {
+		query, err := fs.ReadFile(f.Name)
+		if err != nil {
+			return nil, err
```
здесь тоже ошибку лучше обернуть

dlomanov replied:
обернул ошибку

bambruysk commented on internal/apps/client/ui/components/entry_card_create.go:
```diff
@@ -0,0 +1,271 @@
+}
+
+func (c *EntryCreateCard) Update(msg tea.Msg) (result UpdateResult, cmd tea.Cmd) {
+	if msg, ok := msg.(tea.KeyMsg); ok {
```
слишком длинная функция и больщая вложенность 
Лучше разбить на несколько

каждый кейс можно выделить в отдельную функцию 

вместо послежоватедлльности if лучше использовать switch type

dlomanov replied:
зарефакторил ui-слой

bambruysk commented on internal/apps/client/ui/components/entry_table.go:
```diff
@@ -0,0 +1,257 @@
+			}
+			result.Prev = c.back
+			return result, nil
+		case "j", "k", "up", "down":
```
vim layouy 👍 )

bambruysk commented on internal/apps/server/entities/entry_test.go:
```diff
@@ -0,0 +1,214 @@
+				require.ErrorIs(t, errs, wantErr, "error mismatch")
+				return
+			}
+			if len(tt.wantErrs) == 0 {
+				require.NoError(t, errs, "error mismatch")
+			}
```
в пакете assert есть https://pkg.go.dev/github.com/stretchr/testify/assert#ErrorAssertionFunc

Удобнее использовать его - можно создавать кастомные функции проверки ошибок тоже

dlomanov replied:
заиспользовал в новых тестах, старые тесты перепишу позже

bambruysk commented on internal/apps/server/entrypoints/grpc/interceptor/interceptor.go:
```diff
@@ -0,0 +1,82 @@
+func interceptorLogger(logger *zap.Logger) logging.Logger {
+	return logging.LoggerFunc(func(_ context.Context, lvl logging.Level, msg string, fields ...any) {
+		var values []zap.Field
+		if len(fields)%2 == 0 {
```
а если число полей нечетное то не запишет лое?

bambruysk replied:
Я бы использовал тут map[string]any 
Сложнее накосячить с передаваемыми данными, не нужно приведение типов

dlomanov replied:
переделал на map[string]any, убрал проверку на четность

bambruysk commented on internal/apps/server/usecases/entry.go:
```diff
@@ -0,0 +1,483 @@
+	"go.uber.org/zap"
+)
+
+type (
```
обычно сами типы доменных данных entity выносим в отдельный пакет от бизнес логики
Так как от них зависят все, а зависимостей во вне от бизнес-логики быть не должно

dlomanov replied:
вынес request/response в пакет entity

bambruysk commented on internal/apps/server/usecases/entry_test.go:
```diff
@@ -0,0 +1,283 @@
+	"time"
+)
+
+func TestEntryUC(t *testing.T) {
```
лучше бы разбить на несколько тестов и использовать table-dirven test

dlomanov replied:
Переделал тест, стало удобнее, но заняло много времени. Остальные тесты переделаю в фоне

bambruysk reviewed (approved):
Старался попридираться, но в общем особенно негде. Очень хорошо

dlomanov committed:
support master-password

- sqlite now encrypted with master-password
- added auth by master-password with scrypt
- also added variables for ldflags

dlomanov committed:
PR refactoring

dlomanov committed:
add About-component with build-info

- also added visual enhancements

dlomanov committed:
add entry-type specific components for create/update

dlomanov committed:
add client use-case tests

dlomanov committed:
PR refactoring

dlomanov committed:
add client builds

dlomanov committed:
dockerize server

dlomanov committed:
update cert

dlomanov committed:
use logger in ui

dlomanov committed:
Merge remote-tracking branch 'origin/dev' into dev

dlomanov committed:
update builds

dlomanov committed:
update builds

dlomanov committed:
update builds

dlomanov committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	cmd/client/builds/client_amd64_linux
#	cmd/client/builds/client_amd64_macos
#	cmd/client/builds/client_amd64_windows.exe
#	cmd/client/builds/client_arm64_macos

dlomanov committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	cmd/client/build.sh
#	cmd/client/builds/client_amd64_linux
#	cmd/client/builds/client_amd64_macos
#	cmd/client/builds/client_amd64_windows.exe
#	cmd/client/builds/client_arm64_macos

Commit:
add macos builds

dlomanov committed:
Merge remote-tracking branch 'origin/dev' into dev

dlomanov commented:
- регистрация, аутентификация и авторизация пользователей
      на клиенте — мастер-пароль, которым шифруется sqlite, scrypt
      на сервере — bcrypt, jwt
-  хранение приватных данных
      на клиенте и сервере данные шифруются перед сохранением в БД (AES + GCM)
-  синхронизация данных между несколькими авторизованными клиентами одного владельца
      клиент синхронизируется с сервером в зависимости от действий пользователя
      сервер используется версионирование записей, при конфликте сохраняется оба варианта
-  передача приватных данных владельцу по запросу
      для передачи данных используется gRPC + TLS
-  аутентификация и авторизация пользователей на удалённом сервере
      JWT-токен разворачивается в user ID
      все операции с данными на сервере происходят с использованием user ID
- создание, редактирование и удаление данных на стороне сервера или клиента
      CUD реализовано на стороне клиента с дальнейшей синхронизацией данных с сервером
- формат регистрации нового пользователя
      просто пара login/pass
